### ISSUE TITLE
Shortcuts for under version 7.1

### ISSUE THREAD

**Issue by brianmercer** [brarcher/loyalty-card-locker#154]
Thanks so much for putting in the widget and responding to my review. It does do exactly what I needed by having a widget that goes to a specific card.

I saw that some people are complaining about the cosmetics of the widget and I have that same problem since I have icon text disabled on my launcher pages in Nova Launcher Prime. But I don't mind a cosmetic issue when it is functional.

However, I do think you can solve the cosmetic issues by allowing a "shortcut" to a specific card in lower versions of android. I've used shortcuts in versions 5 and 6 of android and I'm on 6 now. I use the strongswan VPN app that allows shortcuts to a specific profile. Shortcuts would automatically match the text, or have no text if that is set up. Also in Nova you can change launcher icons, so I could manually set a different custom icon for each shortcut.

And finally, if you wanted to publish the intent and data field format of the shortcut here on the github docs then I could have Tasker bring it up automatically whenever I connect to the wifi at the gym which would be very cool.

I'm not sure how difficult it is to enable shortcuts on the lower versions, but I'm hoping it's not a hassle. Thanks again.

**Comment by brarcher** [comment]
Thanks for providing additional information, it is helpful.

My understanding of the feedback from the widget "shortcut" is that the text color may not match that of the other text on the home screen, for example:

<img width="66" alt="screen shot 2017-11-23 at 10 14 36 am" src="https://user-images.githubusercontent.com/5264535/33179527-ba070976-d037-11e7-84fc-f3f6867469c3.png">

where the other text is white. Is it typical to have white text, or could the color be changed via settings on the home screen app or have it change based on the background image color? Changing the default color to white is simple, and maybe it would help most users, but I'm curious if there is a way to have the correct color every time.

> Shortcuts would automatically match the text

What do you mean? The text color always matches the other text on the home screen?

> I use the strongswan VPN app that allows shortcuts to a specific profile.

Does the app allow multiple widget shortcuts, or just one? I found that I was only able to have one unique widget shortcut, and if a second was added it may have different text but it would always launch the same card. Perhaps I missed something when implementing it, or perhaps it was a bug or missing feature in the home app I was testing with.

> if you wanted to publish the intent and data field format of the shortcut here on the github docs

The intent to launch the loyalty card app to display a specific card? The code here [shows](https://github.com/brarcher/loyalty-card-locker/blob/master/app/src/main/java/protect/card_locker/MainActivity.java#L99) what the intent needs to look like. Namely, the Intent is for the LoyaltyCardViewActivity Activity, the action does not matter, and contains a Bundle with two fields:

   id: int
   view: boolean

where "view" should be true and "id" is the id of the card in the database. The ID may be nice internally, as it is unique, but not as friendly externally, as opposed to using a name. You can determine the ID by exporting the cards in CSV format. For example:
```
"_id",store,note,cardid,barcodetype
7,storename 1,,abc123,AZTEC
```

The card named "storename" is ID 7.

There is probably a way to get a specific card to launch which is easier for what you are trying for. If you have ideas, please let me know.

**Comment by brianmercer** [comment]
I'm sorry that I don't know android or java enough to read the code and provide a patch. What you have provided is a perfectly functional "widget", when a "shortcut" would suffice.

I assume a shortcut would be simpler to code since it's just sending the launcher an intent, some data, a default icon and some default title text. Whereas a widget is live and can fetch data and draw itself on a canvas using any images or text. A widget might also use more battery if it is set to update.

Here is a screenshot of the Google Now Launcher on Marshmallow 6.01
![screenshot_20171123-161257](https://user-images.githubusercontent.com/230354/33188978-7aa85b9c-d06c-11e7-9c65-485a3c79a8c4.png)
The widget text is a different color and position and is cut off at the bottom. I believe because the widget has to draw it's own text, while a shortcut is just a text field that is rendered by the Launcher.

The apple, orange, grape, lemon are shortcuts for Connectbot. (https://github.com/connectbot/connectbot) Each shortcut opens a different server profile in Connectbot. I don't think there's a limit.

I will try to use the information provided to make a Tasker task. Thanks.

**Comment by brianmercer** [comment]
I see that "Launcher Shortcuts" are a new 7.0 feature https://developer.android.com/guide/topics/ui/shortcuts.html with more features.

And yet shortcuts themselves go back to API 1. https://developer.android.com/reference/android/content/Intent.html#ACTION_CREATE_SHORTCUT

Sorry I can't be more help. Maybe the code from Connectbot will be helpful.

Thanks again.

**Comment by brarcher** [comment]
I realize now that I missed something really obvious. Instead of finding the API to create a shortcut I tried to make a widget that sort-of did the same thing. Thanks for the links. With that information I'm able to create shortcuts pretty easily:

<img width="132" alt="screen shot 2017-11-24 at 9 14 25 pm" src="https://user-images.githubusercontent.com/5264535/33226470-7d618d4a-d15c-11e7-9640-c8c417ef2092.png">

Here is a question. When would be the best time to ask if a shortcut should be added? I'm thinking when a card is being added or edited there should be a checkbox, asking if a shortcut should be created. If checked, one is created (if it does not already exist).

Given this, I'll be deleting the widget when real shortcut support is added, as the widget missed the point.

**Comment by brarcher** [comment]
This should be resolved in v0.15, released yesterday. If it is not quite working as you would expect or you have additional feedback, please re-open.

**Comment by brianmercer** [comment]
The shortcuts are perfect now.

The only additional feature would be to be able to add the shortcut thru the widget interface of the Launcher.  Same as for the widget, but for the shortcut. I think it is in the manifest:
https://github.com/connectbot/connectbot/blob/3bbf91a08c5dd8234c442df8c5e1a0319276794a/app/src/main/AndroidManifest.xml#L44

and what this guy is explaining here: https://stackoverflow.com/questions/11439253/add-my-app-to-the-add-shortcut-list-to-have-a-shortcut-on-homescreen

You already had the code for selecting a specific card while setting up the widget.

Thanks again.
